### PR TITLE
Distribute shallow water driver.

### PR DESCRIPTION
**Pull request by sriharshakandala**
- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: examples/sphere/shallow_water.jl [hunk 1]
```diff
@@ -105,34 +119,23 @@ else # default case, steady-state test case
...
 quad = Spaces.Quadratures.GLL{Nq}()
-space = Spaces.SpectralElementSpace2D(grid_topology, quad)
-
+if usempi
+    grid_topology = Topologies.DistributedTopology2D(context, mesh)
+    global_grid_topology = Topologies.Topology2D(mesh)
+    space = Spaces.SpectralElementSpace2D(grid_topology, quad)
+    global_space = Spaces.SpectralElementSpace2D(global_grid_topology, quad)
+else
+    comms_ctx = nothing
```
Comment (akshaysridhar) [inline examples/sphere/shallow_water.jl:142]: This block seems to be specific to `Topology` and `Space` assignments, suggest moving the `comms_ctx` variable to the logger block in lines `10-28`. (Is this assignment necessary? I don't see `comms_ctx` being used elsewhere in this example.)

File: examples/sphere/shallow_water.jl [hunk 2]
```diff
@@ -416,106 +435,147 @@ if haskey(ENV, "CI_PERF_SKIP_RUN") # for performance analysis
...
-    )
-    # Height errors over time
-    relL1err = Array{Float64}(undef, div(T, dt))
-    for t in 1:div(T, dt)
-        relL1err[t] = norm(sol.u[t].h .- Y.h, 1) / norm(Y.h, 1)
+if usempi
+    walltime = @elapsed sol = OrdinaryDiffEq.solve!(integrator)
+    ClimaComms.iamroot(context) && println("walltime = $walltime (sec)")
+else
+    sol = @timev OrdinaryDiffEq.solve!(integrator)
```
Comment (akshaysridhar) [inline examples/sphere/shallow_water.jl:442]: This is fine, I'm not sure if this makes a functional difference to keep the expressions in lines `439` and `442` identical in the MPI / serial runs ; Just so I understand the use of `@elapsed` here - are there substantial differences in the evaluations in `@elapsed` and `@timev`, e.g. `gc` inclusion?

File: examples/sphere/shallow_water.jl [hunk 3]
```diff
@@ -416,106 +435,147 @@ if haskey(ENV, "CI_PERF_SKIP_RUN") # for performance analysis
...
+    @info "Fluid volume at time t = 0: ", sum(Y0_global.h)
+    @info "Fluid volume at time t = $(T): ", sum(sol_u_end_global.h)
+
+    if test_name == steady_state_test_name ||
+       test_name == steady_state_compact_test_name
+        # In these cases, we use the IC as the reference exact solution
+        @info "L₁ error at T = $(T): ", norm(solution[end].h .- Y0_global.h, 1)
+        @info "L₂ error at T = $(T): ", norm(solution[end].h .- Y0_global.h)
+        @info "L∞ error at T = $(T): ",
+        norm(solution[end].h .- Y0_global.h, Inf)
```
Comment (akshaysridhar) [inline examples/sphere/shallow_water.jl:487]: Can we include a direct comparison of the serial/ parallel run solution `norm` outputs ?

File: examples/sphere/shallow_water.jl (hunk 1 above)
Comment (sriharshakandala) [reply examples/sphere/shallow_water.jl]: done.

File: examples/sphere/shallow_water.jl (hunk 2 above)
Comment (sriharshakandala) [reply examples/sphere/shallow_water.jl:442]: `@timev` is very informative, but generates garbled output for the distributed case. Therefore, we use `@elapsed` for distributed cases, but still use `@timev` for serial case.

File: examples/sphere/shallow_water.jl (hunk 3 above)
Comment (sriharshakandala) [reply examples/sphere/shallow_water.jl:487]: This is best done outside the driver, possibly in the CI pipeline if needed.

Comment (sriharshakandala) [reply examples/sphere/shallow_water.jl:487]: I will add the corresponding distributed run in the CI pipeline for comparision.

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

File: examples/sphere/shallow_water.jl [hunk 4]
```diff
@@ -416,106 +435,148 @@ if haskey(ENV, "CI_PERF_SKIP_RUN") # for performance analysis
...
-    )
-    linkfig(
-        relpath(
+    @info "Test case: $(test_name)"
+    @info "  with α: $(test_angle_name)"
+    @info "Solution L₂ norm at time t = 0: ", norm(Y0_global.h)
+    @info "Solution L₂ norm at time t = $(T): ", norm(solution[end].h)
+    @info "Fluid volume at time t = 0: ", sum(Y0_global.h)
+    @info "Fluid volume at time t = $(T): ", sum(solution[end].h)
+
```
Comment (akshaysridhar) [inline examples/sphere/shallow_water.jl:481]: Looks like the solution norms for serial vs parallel runs differ by a factor of `0.9999999999999998` i.e. ~ _O_(eps) ; I'll consider this insignificant for our run cases. This isn't expected to change substantially with increasing ranks I hope? (`gather` is elsewhere unit-tested so this is fine with me) LGTM.

**Review by akshaysridhar — APPROVED** [review]

**Comment by sriharshakandala** [comment]
bors try

**Pull request merged**